File: testClients/randomFarm/AI.py
# -*- coding: iso-8859-1 -*-
#-*-python-*-
from BaseAI import BaseAI
from GameObject import *

import random

class mapTile:
  def __init__(self):
    self.gold = 0
    self.pirates = []
    self.port = None

class AI(BaseAI):
  """The class implementing gameplay logic."""
  @staticmethod
  def username():
    return "Shell AI"

  @staticmethod
  def password():
    return "password"

  def init(self):
    pass

  def end(self):
    pass

  def run(self):
    while True:
      print "Turn: ", self.turnNumber()
      myPirates = [i for i in self.pirates if i.getOwner() == self.playerID()]
      map = [ [ mapTile() for i in xrange(self.mapSize()) ] for j in xrange(self.mapSize()) ]
      for i in self.pirates:
        i.gold = i.getGold()
        map[i.getX()][i.getY()].pirates.append(i)
      
      for i in self.ports:
        map[i.getX()][i.getY()].port = i
      
      for i in self.treasures:
          map[i.getX()][i.getY()].gold += i.getGold()
      
      for i in map:
        for j in i:
          if j.pirates and j.pirates[0].getOwner() == self.playerID():
            if j.port:
              for k in j.pirates:
                if k.gold < 20:
                  k.pickupTreasure(20 - k.gold)
                else:
                  k.dropTreasure(k.gold - 20)
                k.gold = 20
            else:
              for k in j.pirates:
                k.dropTreasure(k.gold)
                j.gold += k.gold
                k.gold = 0
              num = len(j.pirates)
              for k in xrange(num):
                j.pirates[k].gold = (j.gold-1)/(num-k)
                j.pirates[k].pickupTreasure(j.pirates[k].gold)
                j.gold -= j.pirates[k].gold
                  
      
      for i in myPirates:
        x, y = random.choice( [(-1, 0), (1, 0), (0, -1), (0, 1)] )
        i.move(i.getX() + x, i.getY() + y)
      
      yield 1
      

  def __init__(self, conn):
      BaseAI.__init__(self, conn)

File: testClients/plantation/AI.cpp
#include "AI.h"
#include "util.h"

AI::AI(Connection* conn) : BaseAI(conn) {}

const char* AI::username()
{
  return "Shell AI";
}

const char* AI::password()
{
  return "password";
}

//This function is run once, before your first turn.
void AI::init()
{
  land.resize(mapSize(), vector<bool>(mapSize(), false));

  for(size_t i=0;i<tiles.size();i++)
  {
    land[tiles[i].x()][tiles[i].y()] = (tiles[i].type()==0);
  }
}

//This function is called each time it is your turn.
//Return true to end your turn, return false to ask the server for updated information.
bool AI::run()
{
  cout<<"Turn Number: "<<turnNumber()<<" My Gold: "<<players[playerID()].gold()<<endl;
  int turningPoint = 100;
  if(turnNumber() > turningPoint+50) return true;
 
  list<Port*> myPorts;
  gold = vector<vector<size_t> >(mapSize(), vector<size_t>(mapSize(),0));
  
  for(size_t i=0;i<ports.size();i++)
  {
    if(ports[i].owner()==playerID())
    {
      myPorts.push_back(&ports[i]);
    }
    land[tiles[i].x()][tiles[i].y()] = false;
  }
  for(size_t i=0;i<pirates.size();i++)
  {
    if(pirates[i].owner()!=playerID())
    {
      land[pirates[i].x()][pirates[i].y()] = false;
    }
  }
  int amount = players[playerID()].gold()/myPorts.size();
  for(list<Port*>::iterator it = myPorts.begin();it!=myPorts.end();it++)
  {
    vector<Pirate*> pickup;
    for(size_t i=0;i<pirates.size();i++)
    {
      if(samePos(pirates[i], **it))
      {
        pickup.push_back(&pirates[i]);
      }
    }
    if(pickup.size()==0)
    {
      /*
      while(pirateCost() < amount)
      {
        (*it)->createPirate();
        amount -= pirateCost();
      }
      */
    }
    else if(turnNumber()< turningPoint)
    {
      amount/=pickup.size();
      for(size_t i=0;i<pickup.size() && amount>0;i++)
      {
        pickup[i]->pickupTreasure(amount);
      }
    }
  }
  // save the amount of gold
  for(size_t i=0;i<treasures.size();i++)
  {
    gold[treasures[i].x()][treasures[i].y()]+=treasures[i].gold();
  }
  list<Tile*> empty, harvest;
  for(size_t i=0;i<tiles.size();i++)
  {
    if(land[tiles[i].x()][tiles[i].y()] && gold[tiles[i].x()][tiles[i].y()]==0)
    {
      empty.push_back(&tiles[i]);
    }
    else if(land[tiles[i].x()][tiles[i].y()])
    {
      harvest.push_back(&tiles[i]);
    }
  }
  for(size_t i=0;i<pirates.size();i++)
  {
    //
    if(pirates[i].owner()==playerID())
    {
      if(players[playerID()].gold()> portCost())
      {
      //  pirates[i].buildPort();
      }
      bool spread = turnNumber()< turningPoint;
      if(!spread)
      {
        vector<MoveHelper<Tile*> > sorted;  
        sortNearest(pirates[i].x(), pirates[i].y(), harvest, sorted, 0);
        if(sorted.size()>0)
        {
          cout<<"Harvesting"<<endl;
          pirates[i].move(sorted[0].path[0]->x(),sorted[0].path[0]->y());
          harvest.erase(sorted[0].it);
        }
        else
        {
          // move to the nearest port
          vector<MoveHelper<Port*> > sorted;
          sortNearest(pirates[i].x(), pirates[i].y(), myPorts, sorted, 0);
          if(sorted.size()>0)
          {
            pirates[i].move(sorted[0].path[0]->x(),sorted[0].path[0]->y());
          }
        }
      }
      if(spread)
      {
        vector<MoveHelper<Tile*> > sorted;  
        sortNearest(pirates[i].x(), pirates[i].y(), empty, sorted, 0);
        // move there
        if(sorted.size()>0)
        {
          pirates[i].move(sorted[0].path[0]->x(),sorted[0].path[0]->y());
          empty.erase(sorted[0].it);
        }
      }
      
      if(at(pirates[i],myPorts))
      {
        if(pirates[i].gold()>0)
        {
          pirates[i].dropTreasure(pirates[i].gold());
        }
      }
      else if(gold[pirates[i].x()][pirates[i].y()]==0 && pirates[i].gold()>0 && spread)
      {
        pirates[i].dropTreasure(1);
        gold[pirates[i].x()][pirates[i].y()]++;
      }
      else if(gold[pirates[i].x()][pirates[i].y()]>1)
      {
        int modifier = 1;
        if(!spread)
        {
          modifier = 0;
        }
        pirates[i].pickupTreasure(gold[pirates[i].x()][pirates[i].y()]-modifier);
        gold[pirates[i].x()][pirates[i].y()]=modifier;
      }
    }
  }
  
  return true;
}

//This function is run once, after your last turn.
void AI::end(){}
